falcon_multiqc/commands/query.py

```diff
@@ -40,156 +40,123 @@
+    select_cols = []
+
+    # Order of the following columns will be the order of output columns.
+    for col in columns:
+        if col == 'sample':
+            select_cols.extend([Sample.id, Sample.sample_name])
+        if col == 'cohort':
+            select_cols.append(Cohort.id)
+        if col == 'batch':
+            select_cols.append(Batch.batch_name)
+        if col == 'tool-metric':
+            select_cols.append(RawData.qc_tool)
+            # If filtering on tool, we need to select for the tool metrics.
+            if tool_metric_filters:
+                for tm in tool_metric_filters:
+                    # Use the metric name as this column's alias.
+                    c = RawData.metrics[tm[1]].label(tm[1])
+                    c.quote = True
+                    select_cols.append(c)
+    if multiqc:
+        # Need to add batch paths.
+        select_cols.append(Batch.path)
 
+    query = Query(select_cols, session=session)
 
-def query_metric(session, query, tool, attribute, operator, value):
-    if not query:
-        return (session.query(Sample.id, Sample.sample_name, Batch.path, Batch.batch_name, Cohort.id, RawData.qc_tool)
-                .join(RawData, Sample.id == RawData.sample_id)
-                .join(Batch, Sample.batch_id == Batch.id)
-                .join(Cohort, Sample.cohort_id == Cohort.id)
-                .filter(RawData.qc_tool == tool, ops[operator](RawData.metrics[attribute].astext.cast(Float), value)))
-    else:
-        return (query.union(session.query(Sample.id, Sample.sample_name, Batch.path, Batch.batch_name, Cohort.id, RawData.qc_tool)
-                            .join(RawData, Sample.id == RawData.sample_id)
-                            .join(Batch, Sample.batch_id == Batch.id).join(Cohort, Sample.cohort_id == Cohort.id)
-                            .filter(RawData.qc_tool == tool, ops[operator](RawData.metrics[attribute].astext.cast(Float), value))))
+    # Add the table joins needed for the given column selection.
+    if 'sample' in columns or 'batch' in columns:
+        # We need Batch for batch.path for both sample and batch.
+        query = query.join(Batch, Sample.batch_id == Batch.id)
+    
+    if 'tool-metric' in columns: 
+        query = query.join(RawData, Sample.id == RawData.sample_id)
+
+    if 'cohort' in columns:
+        query = query.join(Cohort, Sample.cohort_id == Cohort.id)
 
+    return query
+
+# TODO: Currently only supports metrics that are float values. Support more.
+# Returns a sqlalchemy query that queries the database with a filter
+# with the given tool, attribute, operator and value.
+def query_metric(session, query, tool, attribute, operator, value):
+    if operator not in ops:
+        # If user has not given a valid operator, do not filter on metric, just tool.
+        return query.filter(RawData.qc_tool == tool)
+    return query.filter(RawData.qc_tool == tool, ops[operator](RawData.metrics[attribute].astext.cast(Float), value))
```

Comment:
Hey so right now this doesn't work for multiple filters (e.g. tool-metric --tool_metric picard_insertSize MEAN_INSERT_SIZE '>' 565 --tool_metric verifybamid AVG_DP '<' 28 would always return 0 when it should be 36) becuase in the SQL conversion it makes an 'AND' between the two filters. The following it make it so each new filter item gets added as 'OR' 

def query_metric(session, query, tool_metric):
    if operator not in ops:
        # If user has not given a valid operator, do not filter on metric, just tool.
        return query.filter(or_(RawData.qc_tool == tool for tool, attribute, operator, value in tool_metric))
    return query.filter(or_(and_(RawData.qc_tool == tool, ops[operator](RawData.metrics[attribute].astext.cast(Float), value)) for \
            tool, attribute, operator, value in tool_metric))

This generates 36 samples, but it is only 'OR' - i.e. the query can contain either: all filters satisfied, or any combination of filters satisfied 

NOTE: this does not yet solve the problem of getting samples which are both simultaneously the two filters - I will work on the tomorrow or Friday, but that should be doable

Reply:
The SQL looks like:
'WHERE raw_data.qc_tool = %(qc_tool_1)s AND CAST((raw_data.metrics ->> %(metrics_3)s) AS FLOAT) > %(param_1)s OR raw_data.qc_tool = %(qc_tool_2)s AND CAST((raw_data.metrics ->> %(metrics_4)s) AS FLOAT) < %(param_2)s
['sample.id', 'sample.sample_name', 'raw_data.qc_tool', 'raw_data.MEAN_INSERT_SIZE', 'raw_data.AVG_DP']'

which is what we want for now. Next step is to group_by count > x